libukrandom: Introduce driver for CPU-generated entropy + misc. fixes

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): `arm64`, `x86_64`]
 - Platform(s): [N/A]
 - Application(s): [N/A]

### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

This series introduces the driver API of `libukrandom` and introduces a driver for CPU-generated randomness based on the implementation of `ukarch_random`, which is now deprecated.

It also adds a new Kconfing option, namely `LIBUKRANDOM_CMDLINE_INIT` to seed the CSPRNG via the cmdline. In ChaCha this is used to pass 8x 32-bit words for the key in `random.seed` cmdline paramter. When this option is enabled AND initialziation paramters are passed to the cmdline, the underlying RNG driver is not used. This replaces `CONFIG_LIBUKRANDOM_SEED_INSECURE` as it allows users of legacy hardware or older QEMU versions to seed the CSPRNG from the host's entropy sources.

Finally, it additionally relaxes the requirement of `libukrandom` in the config of KVM to avoid compatibility issues with legacy hardware and versions of QEMU < 8 on applications that don't require an RNG.

michpappas committed:
drivers: Add skeleton for the ukrandom subsystem

Add skeleton for the ukrandom driver subsystem. This will host drivers
that implement the libukrandom driver API.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
arch/arm64: Generate MTE keys with libukrandom

Update MTE to generate keys using the uk_random API.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/ukrandom: Introduce driver for CPU generated randomness

This is based on the implementation of ukarch_random (RDRAND / RDSEED
for x86_64, RNDR / RNDRRS for arm64).

Register to earlytab after earlycon. Libraries that require CSRNG
facilities at earlytab must register at a priority higher than
UK_RANDOM_EARLY_DRIVER_PRIO.

On arm64 this requires a CPU revision of at least armv8-5.a that
implements FEAT_RNG.

On x86_64 the minimum CPU requirement is Intel Ivy Bridge / AMD
Excavator (2012). RDSEED support is now optional, as it was introduced
in Intel Broadwell / AMD Zen (2015), so CPUs released between 2012 and
2015 support RDRAND but not RDSEED. On these CPUs seed_bytes() will
return an error, and seed_bytes_fb() will fall-back to random_bytes().

Checkpatch-Ignore: LONG_LINE
Checkpatch-Ignore: OPEN_ENDED_LINE
Checkpatch-Ignore: EMBEDDED_FUNCTION_NAME
Checkpatch-Ignore: SPACING

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukrandom: Update library to use the driver API

Update libukrandom to use the newly defined driver API. Make sure that
boot does not fail if libukrandom has not been successfully initialized.
Runtime errors should be returned instead. Specifically, getrandom() will
return -ENOSYS and device nodes won't be created. Similarly, calls to
the kernel API will return a negative value.

Update ChaCha to use seed with fallback, to handles cases where the
conditioned entropy is exhausted. Update /dev/hwrng to fill the buffer
with random_bytes() instead of random_seed().

Checkpatch-Ignore: AVOID_EXTERNS
Checkpatch-Ignore: ENOSYS
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukrandom: Define the driver API

Define the driver API of libukrandom. This is a generic API that
deprecates the CPU-oriented ukarch_random API.

Checkpatch-Ignore: SPACING
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukrandom: Add initialization via the cmdline

Add a new Kconfing option, namely LIBUKRANDOM_CMDLINE_INIT
to seed the CSPRNG via the cmdline. In ChaCha this is used
to pass 8 bytes for the key followed by 2 bytes for the nonce
in `random.seed` cmdline parameter.

When this option is enabled AND initialziation parameters are
passed to the cmdline, the underlying RNG driver is not used.

lib/ukrandom: Seed the CSPRNG from the cmdline

Provide the ability to seed ChaCha20 from the kernel's command
line. This is controlled by a newly introduced Kconfig option,
CONFIG_LIBUKRANDOM_CMDLINE_SEED. The relevant cmdline parameter
is `random.seed` that should by populted by the loader with an
array of 8 random 32-bit unsigned integers for the ChaCha20 key.

Seeding via the cmdline takes precedence over the device-tree and
drivers to RNG devices.

Checkpatch-Ignore: SPACING
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/kvm: Relax requirement for RDRAND

Enabling cpu-generated randomness by default breaks execution on
legacy x86_64 CPUs that don't support RDRAND, as well as older
versions of QEMU provided by some distros when executing with TCG,
given that RDRAND support in TCG is introduced in QEMU 8.0.0.

Update config to enable CPU-generated randomness conditionally
to the selection of libukrandom. The TRNG will only be enabled
by applications that select libukrandom.

While this does not fix the problem at its core, it reduces its
scope to applications that explicitly require randomness.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukrandom: Remove redundant IRQ ASSERTION check

This assertion is redundant and breaks earlytab calls that
happen before we configure the interrupt controller. Protect
the global context with a spinlock instead.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/common: Generate PAuth keys with libukrandom

Update PAuth to generate keys using the uk_random API.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

nderjung commented:
One usability request to simplify seeding the driver would be to allow for using a simple numeric, such that the following is possible:

```
random.seed=12345678
```

vs.

```
random.seed=[ 0xb0 0xb0 0xf0 0x0d 0xb0 0xb0 0xf0 0x0d 0xd0 0xd0 ]
```

Alex-deVis commented on lib/ukrandom/chacha.c:
```diff
@@ -35,7 +35,7 @@
 #include <uk/assert.h>
 #include <uk/ctors.h>
 #include <uk/random.h>
-#include <uk/arch/random.h>
```
This file was deleted in the previous commit, are we OK with intermediary commits that do not compile?

michpappas replied:
Yes that's fine, as long as the PR compiles, as the amount of work it often takes to make every commit to compile is disproportional to the benefit.

skuenzer replied:
However, this is not excluding that it is in general a good practice to keep compile-able individual commits. It helps to structure PRs and thinking step-wise. However, we relaxed the requirement because for bigger re-arch PRs, it can be in fact difficult to achieve that.

mariapana reviewed (approved):
Looks good to me.

Reviewed-by: Maria Pana <maria.pana4@gmail.com>

Alex-deVis commented on drivers/ukrandom/lcpu/arch/arm64/random.c:
```diff
@@ -0,0 +1,128 @@
+	 * of RNDRRS (DDI 0487K.a Sect. DD23.2.148)
+	 */
+	rc = seed_bytes(buf, size);
+	if (unlikely(rc))
```
I think adding a debug message here would be helpful, as the failure of `seed_bytes` is a hint that the CPU does not provide enough true randomness for the app.

michpappas replied:
I'd like to keep the very low level code minimal. As the driver ops are protected with the `__check_result`, the caller (`lib/ukrandom/chacha.c`) should handle this and print an appropriate message (it does).

Alex-deVis commented on drivers/ukrandom/lcpu/arch/x86_64/random.c:
```diff
@@ -0,0 +1,149 @@
+
+	ukarch_x86_cpuid(1, 0, &eax, &ebx, &ecx, &edx);
+	if (unlikely(!(ecx & X86_CPUID1_ECX_RDRAND))) {
+		uk_pr_err("CPU does not implement RDRAND\n");
```
I think this should be "The CPU does not ..." to match the RDSEED warn message.

michpappas replied:
Ok

Alex-deVis commented on drivers/ukrandom/lcpu/arch/x86_64/random.c:
```diff
@@ -0,0 +1,149 @@
+	__u8 res;
+
+	if (!have_rdseed)
+		return -ENOTSUP;
```
The commit message states that **-ENODEV** is returned if RDSEED is not supported. I think the `-ENOTSUP` is more appropriate and the commit message should be updated.

michpappas replied:
Good catch, thanks 👍🏼

Alex-deVis reviewed (commented):
This looks very nice, I left a few inline comments.

Besides, there is a typo in the **drivers/ukrandom: Introduce driver for CPU generated randomness** commit's description: "On CPUs that don't support RDSEED init will no **longer** fail ...".

michpappas committed:
lib/ukrandom: Limit libukrandom seed to ChaCha key

Unlike encryption, the nonce is not important when generating random
numbers as ChaCha20 can generate 2**64 blocks / 1ZiB with the same
(key, nonce) pair.

Set the nonce to a fixed value similarly to other ChaCha-based CSPRNG
implementations like [1]. This brings us in line with the device-tree's
`/chosen/rng-seed` node used by linux that consists of eight 32-bit
unsigned random integers for initializing the ChaCha key, allowing
Unikraft to use unmodified dtbs generated by VMMs.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukrandom: Add Kconfig option for getrandom()

Add a Kconfig option to enable the getrandom() syscall conditionally
to the syscall shim, to enforce library dependencies.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukrandom: Seed the CSPRNG from the device-tree

Provide the ability to seed ChaCha20 from the `/chosen/rng-seed` node
of the dtb. This is controlled by a newly introduced Kconfig option
CONFIG_LIBUKRANDOM_DTB_SEED, enabled by default in platforms that
select LIBFDT.

Seeding via the device-tree takes precedence over the drivers, so if
a seed is set in the dtb, any driver selected is not used.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukofw: Add helper for /chosen/rng-seed

Add fdt_chosen_rng_seed() to populate pointers with the address
of the seed buffer and seed length.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/xen: Relax requirement for RDRAND

Enabling cpu-generated randomness by default breaks execution on
legacy x86_64 CPUs that don't support RDRAND. Update config to
enable CPU-generated randomness conditionally to the selection
of libukrandom.

In x86_64 libukrandom will seed the CSPRNG using RDRAND if the CPU
implements these instruction, unless LIBUKRANDOM_CMDLINE_SEED is
selected in the config.

On arm64 libukrandom will check if a device-tree is provided and
if a seed is set in the `/chosen/rng-seed` property, and otherwise
fall back to RNDR / RNDRRS if the CPU implements FEAT_RNG.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukrandom: Add testsuite for ChaCha20

Add testsuite for ChaCha20. This implements the test vectors defined
in RFC8439: https://datatracker.ietf.org/doc/html/rfc8439

Checkpatch-Ignore: REPEATED_WORD
Checkpatch-Ignore: AVOID_EXTERNS
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukrandom: Increase ChaCha rounds to 20

Although 8 rounds most likely provide enough security [1], ChaCha20
brings us in alignment with Linux, and allows us to implement tests
using the vectors defined in RFC8439.

[1] https://eprint.iacr.org/2019/1492.pdf

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas commented:
Added unit tests for ChaCha20

Alex-deVis reviewed (approved):
Reviewed-by: Alex Apostolescu <alexx.apostolescu@gmail.com>

razvand commented on lib/ukrandom/dev.c:
```diff
@@ -79,13 +70,17 @@ static int dev_random_read(struct device *dev __unused, struct uio *uio,
-	uk_random_fill_buffer(buf, count);
+	ret = uk_random_fill_buffer(buf, count);
+	if (unlikely(ret))
+		return rc;
```
```suggestion
		return ret;
```

Otherwise, this gives a build error. If you enable `LIBUKRANDOM_DEVFS`.

michpappas replied:
Thanks, fixed

razvand reviewed (changes requested):
@michpappas , see the build error after enabling `CONFIG_LIBUKRANDOM_DEVFS`.

For a Python application build (such as the one [here](https://github.com/unikraft/catalog-core/tree/main/python3-hello)), I need to enable `CONFIG_LIBUKRANDOM_GETRANDOM` for it to work (as part of the `Config.uk` file of the application). Otherwise, the application cannot generate random numbers:

```text
Fatal Python error: _Py_HashRandomization_Init: failed to get random numbers to initialize Python
Python runtime state: preinitialized
```

Shouldn't this configuration (`CONFIG_LIBUKRANDOM` and `CONFIG_LIBUKRANDOM_GETRANDOOM`) be part of the `Config.uk` file of the [`lib-python3` library](https://github.com/unikraft/lib-python3)? Because it should be enabled irrespective of the application. Any time you use Python3, this is going to be required.

michpappas commented:
> @michpappas , see the build error after enabling `CONFIG_LIBUKRANDOM_DEVFS`.
> 
> For a Python application build (such as the one [here](https://github.com/unikraft/catalog-core/tree/main/python3-hello)), I need to enable `CONFIG_LIBUKRANDOM_GETRANDOM` for it to work (as part of the `Config.uk` file of the application). Otherwise, the application cannot generate random numbers:
> 
> ```
> Fatal Python error: _Py_HashRandomization_Init: failed to get random numbers to initialize Python
> Python runtime state: preinitialized
> ```
> 
> Shouldn't this configuration (`CONFIG_LIBUKRANDOM` and `CONFIG_LIBUKRANDOM_GETRANDOOM`) be part of the `Config.uk` file of the [`lib-python3` library](https://github.com/unikraft/lib-python3)? Because it should be enabled irrespective of the application. Any time you use Python3, this is going to be required.

@razvand that's right, we need to identify which libraries / applications require secure randomness and enable `CONFIG_LIBUKRANDOM` there. Moreover, libraries that require the `getrandom()` syscall need to enable `CONFIG_LIBSYSCALL_SHIM` and `CONFIG_LIBUKRANDOM_GETRANDOM`. Libraries that require device nodes need to enable `CONFIG_LIBVFSCORE` and `LIBUKRANDOM_DEVFS`.

razvand commented:
> @razvand that's right, we need to identify which libraries / applications require secure randomness and enable `CONFIG_LIBUKRANDOM` there. Moreover, libraries that require the `getrandom()` syscall need to enable `CONFIG_LIBSYSCALL_SHIM` and `CONFIG_LIBUKRANDOM_GETRANDOM`. Libraries that require device nodes need to enable `CONFIG_LIBVFSCORE` and `LIBUKRANDOM_DEVFS`.

@michpappas , roger. Taking a look.

skuenzer commented on drivers/ukrandom/lcpu/arch/x86_64/random.c:
```diff
@@ -0,0 +1,149 @@
+
+	ukarch_x86_cpuid(1, 0, &eax, &ebx, &ecx, &edx);
+	if (unlikely(!(ecx & X86_CPUID1_ECX_RDRAND))) {
+		uk_pr_err("The CPU does not implement RDRAND\n");
```
Hum, should we really say `"implement"`? On x86 you are able to mask out and switch off CPU features for guests (via cpuid), although the CPU provides the feature in hardware. This is often done to allow live migration between different x86 CPU generations by allowing only a common feature set.

What would you think of:
`Hardware random number generation (RDRAND) not available on this CPU`

michpappas replied:
That's true, but IMO it also depends on the perspective. That is, whether virtualized or not, from the operating system's point of view the CPU still appears not to implement the feature. Besides, normally the assumption is that the system runs natively, and virtualization is a special case. In fact, even there, (ideally) the OS is not aware that it's executing over virtualization. So in conclusion I think that this change is unnecessary. If you have a strong opinion though, I can change it.

skuenzer replied:
Hum, you could even argue that on a firmware level (e.g., BIOS) such features could be switched off although present on the CPU. Because of this, I'd prefer to name it `"not available"`. The thought behind this is that it does not lead to any conclusion if the CPU is in principle able to do it or not.

michpappas replied:
Done

skuenzer commented on drivers/ukrandom/lcpu/arch/x86_64/random.c:
```diff
@@ -0,0 +1,149 @@
+	if (ebx & X86_CPUID7_EBX_RDSEED)
+		have_rdseed = __true;
+	else
+		uk_pr_warn("The CPU does not implement RDSEED");
```
Suggestion:
`Hardware random seed generation (RDSEED) not available on this CPU, using RDRAND`

michpappas replied:
Adding "using RDRAND" for now, for the rest see https://github.com/unikraft/unikraft/pull/1496#discussion_r1858301885.

skuenzer replied:
See my answer of that comment, too :+1:

skuenzer commented on lib/ukrandom/Config.uk:
```diff
@@ -1,25 +1,23 @@
+		seed from the cmdline, if the `random.seed` parameter is
+		passed with a 256-bit key formatted as an array of
+		32-bit words. This option has higher precedence than the
+		device-tree and drivers to hardware RNG devices.
```
Should we maybe write one sentence that the cmdline parameter user is responsible for securing access the the command line origin (e.g., boot loader)?

skuenzer commented on drivers/ukrandom/lcpu/Config.uk:
```diff
@@ -0,0 +1,11 @@
+menuconfig LIBUKRANDOM_LCPU
+	bool "CPU generated randomness"
+	depends on (ARCH_X86_64 || (ARCH_ARM_64))
+	select HAVE_ARM64_FEAT_RNG if ARCH_ARM_64
```
I am not fully understanding that we select `HAVE_ARM64_FEAT_RNG` (setting to true) here and at the same time in the Arm CPU configuration. Shouldn't the random driver depend on `HAVE_ARM64_FEAT_RNG`?

michpappas replied:
So in the Arm CPU configuration  `HAVE_ARM64_FEAT_RNG` is an invisible option only, that is also selected by features that depend on randomness. It needs to be there because we use it to generate GCC flags: https://github.com/unikraft/unikraft/blob/staging/arch/arm/arm64/Makefile.uk#L68. Ideally I would had liked that these flags are generated per-library, but unfortunately there are some inter-dependencies on the `march` flag that we need to resolve first at the build system. Example: 

* `FEAT_RNG` resolves to ARCH_REV := armv8.5-a and ARCH_OPT := $(ARCH_OPT)+rng
* `FEAT_PAUTH` resolves to ARCH_REV := armv8.3-a and BRANCH_PROTECTION := pac-ret+leaf
 
At the final compiler flags we need to select the highest architectural revision, otherwise we'll get a compiler error that `FEAT_RNG` is not supported in `armv8.3-a`. For now I choose to have libraries select the architectural features they support (e.g. `FEAT_RNG`) and delegate the compiler flags generation to a single Makefile where we easily can do the dependency resolution by specifying features of later architectural revisions last, so that they override earlier ones.

> Shouldn't the random driver depend on HAVE_ARM64_FEAT_RNG?

The reason we can't have `libukrandom` depend on `FEAT_RNG` is that we don't know if the underlying CPU implements `FEAT_RNG`. On `arm64` GCC defines some generic CPUs , i.e. `MCPU_ARM64_NONE`, `MCPU_ARM64_NATIVE` and `MCPU_ARM64_GENERIC`, and the default we use in Kconfig is`MCPU_ARM64_NONE`. So when the user selects the lcpu rng driver of `arm64`, it is expected that the user implies that the CPU implements `FEAT_RNG`.

skuenzer replied:
Okay, understand. As we discussed, let's re-visit this integration when we touch the compiler settings as part of the final re-arch of the architecture/lcpu libraries. I am thinking that we may want to rename `HAVE_ARM64_FEAT_RNG` to `NEED_ARM64_FEAT_RNG` or `WANT_ARM64_FEAT_RNG` - but getting to a conclusion is not trivial at this point and needs further discussion. We shouldn't block this PR because of this. Can you add a comment to the source code that describes why we are doing this here (for now) with `HAVE_ARM64_FEAT_RNG`?

michpappas replied:
Done

skuenzer commented on drivers/ukrandom/lcpu/init.c:
```diff
@@ -0,0 +1,44 @@
+	return 0;
+}
+
+#if CONFIG_LIBUKBOOT
```
Hum, is it actually possible that we do not need to compile this file at all if `CONFIG_LIBUKBOOT` is absent (via `Makefile.uk`)? I expect that we will see an unused warning for `uk_random_lcpu_init` if `libukboot` is not part of the build (this is theoretically for now of course).

michpappas replied:
The motivation was to leave `init()` available for platforms that may not wish to use `libukboot`, so that they can still initialize the library by calling `init()` directly. In that case, we can address the warning by adding `__maybe_unused`. Overall I don't have a strong opinion, as we will need to formalize the behavior of library initialization in the scope of `libukboot` rearch anyway.

skuenzer replied:
Okay, in such a case `uk_random_lcpu_init()` should not be declared static and we should provide it via `exportsyms.uk`. Secondly, can't we ignore the input parameter and just declare it with unspecified parameters?: `int uk_random_lcpu_init()`. This would look more compatible to non-`libukboot` builds. Additionally we would need to provide the function prototype with a driver-specific header, like `uk/ukrandom/lcpu/lcpu.h`. We can then decide if we make it only available for non-`libukboot` builds which I guess would make sense.

michpappas replied:
Okay, since we don't have a final decision on the behavior of drivers when it comes to that aspect I'll follow your initial suggestion and make compilation of that file conditional to libukboot.

skuenzer commented on arch/arm/arm64/memtag.c:
```diff
@@ -72,11 +72,11 @@ int ukarch_memtag_init(void)
-		UK_CRASH("Could not generate MTE seed\n");
+	rc = uk_random_fill_buffer(&seed, sizeof(seed));
+	if (unlikely(rc)) {
+		uk_pr_err("Could not generate MTE key (%d)\n", rc);
```
Somehow I am tending to use `"Failed to generate..."` instead of `"Could not generate..."` since it is underlining that there is a serious problem. However, we are talking here probably about preference.
If we fail, what is the actual effect? Are we able to boot or do we fail if we do not have MTE?

michpappas replied:
> Somehow I am tending to use "Failed to generate..." instead of "Could not generate..." since it is underlining that there is a serious problem. However, we are talking here probably about preference.

Indeed, I think this is subjective :) Saying that, the idea is that the library does not enforce policy, so the severity of a failure is decided by the caller. In our case we treat failures of initcalls as critical.

> If we fail, what is the actual effect? Are we able to boot or do we fail if we do not have MTE?

Right now we call `UK_CRASH()`. The same will happen when we move MTE initialization to `earlytab`. IMO, if I had to choose between continuing boot with a warning or crashing, I would favor security and choose the latter, given that the fact that someone enabled MTE explicitly in Kconfig implies that they care about that feature. Nevertheless, we can also consider adding a feature in `libukboot` or individual libraries like `ukmemtag` that allows the user to select the system behavior upon failure.

skuenzer commented on plat/common/arm/pauth.c:
```diff
@@ -72,27 +28,40 @@ static inline bool pauth_supported(void)
 
-	if (unlikely(!pauth_supported()))
+	if (unlikely(!pauth_supported())) {
+		uk_pr_err("The CPU does not implement FEAT_PAUTH\n");
```
I have a similar comment here regarding the message as I mentioned to the x86 `RDRAND` feature regarding "implement". Is it possible that features can be masked out on Arm for virtualized guests?

michpappas replied:
Yes, on `arm64` the hypervisor can trap the `mrs` to the feature identification registers and return a custom value to the guest. For the rest let's continue the discussion in https://github.com/unikraft/unikraft/pull/1496#discussion_r1858301885.

skuenzer replied:
Okay, then the same answer in that comment applies here 😄

skuenzer commented on plat/common/arm/pauth.c:
```diff
@@ -72,27 +28,40 @@ static inline bool pauth_supported(void)
-	ukplat_pauth_gen_key(&key_hi, &key_lo);
+	rc = uk_random_fill_buffer(&key_hi, sizeof(key_hi));
+	if (unlikely(rc)) {
+		uk_pr_err("Could not generate APIAKey (%d)\n", rc);
```
Also here and the message at line 62, should we say `"Failed to ..."`? Again probably preference and I am not strongly object `"Could not..."`.

michpappas replied:
See https://github.com/unikraft/unikraft/pull/1496#discussion_r1858175667

skuenzer reviewed (approved):
Thank you so much for this addition!

Reviewed-by: Simon Kuenzer <simon@unikraft.io>

michpappas committed:
lib/ukrandom: Tighten the implementation of ukswrand

Add an internal check to ensure that the CSPRNG has been initialized.
This is an additional check so that we don't merely rely on the caller's
logic. Use __check_result on the definition of uk_swrand_randr().

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

skuenzer reviewed (approved):
Updates are looking good! Thanks!

Reviewed-by: Simon Kuenzer <simon@unikraft.io>
Approved-by: Simon Kuenzer <simon@unikraft.io>
